x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java

```diff
@@ -140,6 +195,108 @@ public <T extends TransportResponse> void sendRequest(
+                    && false == User.isInternal(effectiveSubject.getUser())
+                    && Arrays.stream(effectiveSubject.getUser().roles()).noneMatch(ReservedRolesStore::isReserved);
+            }
+
+            private <T extends TransportResponse> void sendWithRemoteAccessHeaders(
+                final Transport.Connection connection,
+                final String action,
+                final TransportRequest request,
+                final TransportRequestOptions options,
+                final TransportResponseHandler<T> handler
+            ) {
+                logger.debug("Sending request with remote access headers for [{}] action", action);
+
+                if (connection.getVersion().before(VERSION_REMOTE_ACCESS_HEADERS)) {
+                    handler.handleException(
+                        new TransportException(
+                            "Settings for remote cluster indicate remote access headers should be sent but target cluster version ["
+                                + connection.getVersion()
+                                + "] does not support receiving them"
+                        )
+                    );
+                    return;
+                }
+
+                final Authentication authentication = securityContext.getAuthentication();
+                assert authentication != null : "authentication must be present in security context";
+
+                final Optional<String> remoteClusterAlias = remoteClusterAliasResolver.apply(connection);
+                assert remoteClusterAlias.isPresent() : "remote cluster alias must be set for the transport connection";
+
+                final ThreadContext threadContext = securityContext.getThreadContext();
+                final var contextRestoreHandler = new ContextRestoreResponseHandler<>(threadContext.newRestorableContext(true), handler);
+                authzService.retrieveRemoteAccessRoleDescriptorsIntersection(
+                    remoteClusterAlias.get(),
+                    authentication.getEffectiveSubject(),
+                    ActionListener.wrap(roleDescriptorsIntersection -> {
+                        try (ThreadContext.StoredContext ignore = threadContext.stashContext()) {
+                            writeCredentialForClusterToContext(remoteClusterAlias.get(), threadContext);
+                            new RemoteAccessAuthentication(authentication, roleDescriptorsIntersection).writeToContext(threadContext);
+                            sender.sendRequest(connection, action, request, options, contextRestoreHandler);
+                        }
+                    }, e -> contextRestoreHandler.handleException(new SendRequestTransportException(connection.getNode(), action, e)))
+                );
+            }
+
+            private boolean isWhitelistedForRemoteAccessHeaders(final TransportRequest request) {
+                return request instanceof ShardSearchRequest
```

Comment:
I think the action names in org.elasticsearch.action.search.SearchTransportService may be the definitive list of actions we want to allow.

Reply:
Thanks for the pointer! I'm seeing a few more candidate request types there (particularly around scroll). Some of the corresponding classes are package private, so I'm considering switching the allowlist to be based on action name instead of request type. The list will be longer (many actions map to the same request type) but this approach may more conventional and explicit. Still 👀 this one, and working out how scroll contexts work...

Reply:
> Using action names is likely the way to go.

Yup! I will push a preliminary list of action names to this PR and create an issue to track finalizing it. I'll go with a static set of all relevant actions which should give us a fast membership check.

> More specifically, all of them use the _system user to collect remote nodes (via GetClusterState cluster:monitor/state). Async search uses _async_search user to read and store partial results.

Absolutely. In sniff mode, even "regular" cross cluster search involves a GetClusterState action by the system user.

I agree that both support for internal users and the correct list of action names are high priority.

I like your suggested approach around handling internal users. There are a few caveats I want to think through here: 

I don't think we want to grant *all* access that a system user has within a cluster to remote cluster calls, and instead limit it to only the necessary actions (for the system user and barebones CCS it's only `cluster:monitor/state`). For example, the system user has `indices:data/write/*` access which we don't want to enable across clusters -- a malicious admin (or anyone else with access to the remote access API key credential) on a querying cluster could otherwise issue manually-crafted requests with a system user authentication instance in the header and escalate privileges.

There are three ways to go about this that I can think of: 

1. introduce new internal users (e.g. `remote_access_system_user`)
2. handle calls from existing internal users differently on the fulfilling cluster side, when processing a remote access request
3. enforce a privilege intersection with the remote access API key, and instruct administrators to manually include the privileges they require (for example, if an admin does not want to support async search, they don't have to include async user related privileges). This last option has more burden on the end-user but also gives them more control, and makes them explicitly aware of the full access scope a remote cluster has. As such, I don't want to dismiss this option just yet.

I'll put my thinking around the above in a Jira issue and get feedback there. Thanks for raising this, spot on as always!